### PR TITLE
Tile background and ui tweaks

### PR DESCRIPTION
**Pull request by heinoapt**
Includes updated events and images for events, size thumbnail 200x200, medium 500x500 and the original file.

### CODE REVIEW

**Comment by antti-TIEA4** [comment]
Looks good and works. I'm just thinking if there's a way to do this without the directive. Did you try to do this without it first?

**Comment by heinoapt** [comment]
I went straight for the directive as it seems intuitive for me. Other option that comes to my mind is inline CSS, but that isn't good. Do you have other ideas?

**Comment by antti-TIEA4** [comment]
Does `md-card` have an element for background image?

Inline CSS looks ugly but it would be more simple. Though the directive is not that complex (in lines) but it does add another watcher/observer.

Also how about adding the image url to the events json file as a property? Or is there some drawback I'm missing?

**Comment by heinoapt** [comment]
You mean a ready directive or an element inside the md-card? Not that I know of.

I thought it wouldn't be useful to have the URLs in the JSON for every image as I only added one image for each event. If it's upgraded to multiple different images then the URL or other marker is needed. Simplicity was what I was looking for too.

**Comment by antti-TIEA4** [comment]
Btw, the images you found are really nice!
I'm no web developer so this might be a stupid question but do we need multiple resolution versions of the images? Can't we handle the resizing via CSS or other styling?

`md-card` has `` but after quick testing I couldn't make it in the background. Also searched for "md-card background image" but didn't find much.

File: app/index.html [hunk 1]
```diff
@@ -24,6 +24,8 @@
   <script type="text/javascript" src="js/livebus/livebus.factory.js"></script>
 
   <script type="text/javascript" src="js/shared/shared.module.js"></script>
+  <script type="text/javascript" src="js/shared/ucBackground/ucBackground.module.js"></script>
+  <script type="text/javascript" src="js/shared/ucBackground/ucBackground.directive.js"></script>
```
Comment (antti-TIEA4) [inline app/index.html]: As this is a directive for the events' background I think it should go into the `event` folder and called something like `event.background.directive` etc.

File: app/js/shared/ucBackground/ucBackground.directive.js [hunk 2]
```diff
@@ -0,0 +1,13 @@
+'use strict';
```
Comment (antti-TIEA4) [inline app/js/shared/ucBackground/ucBackground.directive.js]: The contents should be wrapped in an anonymous function to protect it from namespace collisions and minification. See the styleguide: https://github.com/johnpapa/angular-styleguide/blob/master/a1/README.md#iife

All the .js files have this so you can just copy paste the first and last lines from any .js file inside the app directory

File: app/index.html [hunk 3]
```diff
@@ -49,7 +51,7 @@
 
     <div ng-if=$root.fullscreen flex layout-fill layout="row">
       <md-sidenav md-is-open="true" md-component-id="left" ng-click="ec.toggleEventList()" class="md-sidenav-left md-whiteframe-z3">
-        <md-card ng-repeat="event in ec.events | orderBy : 'distanceToUser'" ng-cloak flex md-ink-ripple ng-click="ec.selectEvent(event)" ng-class="{'selected' : event === event.selectedEvent }" class="event-list disable-text-selection">
+        <md-card ng-repeat="event in ec.events | orderBy : 'distanceToUser'" ng-cloak flex md-ink-ripple ng-click="ec.selectEvent(event)" ng-class="{'selected' : event === event.selectedEvent }" class="event-list disable-text-selection" uc-background="http://abc.png">
```
Comment (antti-TIEA4) [inline app/index.html]: Did the event list break if you used one of the images here? :smile: 
Or did you just forget this `"http://abc.png"` here?

Comment (heinoapt) [inline app/index.html]: Totally went unnoticed! I deleted it.

File: app/index.html (hunk 1 above)
Comment (heinoapt) [inline app/index.html]: That's true, but a similar function could be used anywhere. That's why added it to shared. You don't think it's needed? The size styling should be somewhere else now that  I think of it.

File: app/js/shared/ucBackground/ucBackground.directive.js (hunk 2 above)
Comment (heinoapt) [inline app/js/shared/ucBackground/ucBackground.directive.js]: Ok, that didn't come to my mind. I thought using `var` was a way to tackle that, but minifying changes that too.

File: app/index.html (hunk 1 above)
Comment (heinoapt) [reply app/index.html]: Now you can use that single directive to any element and style it independently. That makes it worthwhile to have at the `shared/`, what do you think?

Comment (antti-TIEA4) [inline app/index.html]: We used to have all similar components in their own folders (i.e. directives would have been in one place) but then we changed according to the Angular style guide (https://github.com/johnpapa/angular-styleguide/blob/master/a1/README.md#naming) so that related components are in their respective feature folders. I don't think we need to change the CSS background for anything else than the events, so it should be in the events. And if the need arises then we can generalize it. If we generalized all components at first then everything would be in the same place :)

Comment (heinoapt) [reply app/index.html]: The feature based directories make sense, but that is a rather different case in my opinion. Having all the code in a single folder would clutter fast. It's still a style of organizing code, just a different approach to having for example controllers in their own folder. Both have their merits, but this really isn't about extracting directives to their own folder at all.

The background directive being in it's own module means that you could simply copy it to a different project if needed. The shared methods don't need to be restricted to a single app.

What negatives does the module have in your opinion?

Comment (heinoapt) [reply app/index.html]: BTW the multiple resolution files were in case a larger file wasn't needed. Large downloads should be avoided on mobile devices, but CSS sizing works well too.

Comment (antti-TIEA4) [inline app/index.html]: If it's meant to be a general module like that then it should drop the `uc` prefix from all the names. And perhaps `backgroundImage` for the module name, `background.image.module.js` for the module file name and `background.image.directive.js` for the directive.

And then I would still place the module under the `event` folder and inject it only there because it is not yet used by anywhere else.

Comment (heinoapt) [reply app/index.html]: The prefix was from the style guide https://angular.io/styleguide#!#02-07
Quote:
- Why? Prevents element name collisions with components in other apps and with native HTML elements.
- Why? Makes it easier to promote and share the component in other apps.
- Why? Components are easy to identify in the DOM.

We have discussed this at length already and I'm prepared to change my stance on this. I don't really see using modular design being needed only when a component is already needed in other places. Especially if it's this easy to achieve beforehand. Seems like our design principles are just a bit different on this?

Comment (antti-TIEA4) [inline app/index.html]: That part is specifically for components (Angular 1.5 feature) which we are not using but below that there's a guide for directives. They seem to prefer to use a prefix for the directive selector name. I'm ok with that but I don't know if that applies for general directives supposed to be shared by various projects? The filenames and modules don't seem to have the app prefix in that style guide.

Comment (heinoapt) [reply app/index.html]: Well true, that's for Angular2. I don't think the first point is invalid for Angular 1 either?

The naming convention doesn't work exactly like that, I thought the suffix was meant to be the type for the class, not for replacing camel case?

Comment (heinoapt) [reply app/index.html]: What I meant with modular design wasn't really what Angular defines as modules. I do understand the way they are organized.

But good to hear we aren't that far off! If this was organized differently, just a few lines of code would be different. After minimizing the requests won't increase and the code isn't cluttered. I don't really see negatives.

Comment (antti-TIEA4) [inline app/index.html]: The directive name can (and you are right should) have that `uc` prefix.
The module name and filenames shouldn't have it. These comments are below the file imports and at some part I probably forgot what the actual directive content looked like.

So `backgroundImage` for the module name, `background.image.module.js` for the module file name and `background.image.directive.js` or something like that where there's no reference to the app name.

Comment (heinoapt) [reply app/index.html]: It was that which I gave a response above:
"The naming convention doesn't work exactly like that, I thought the suffix was meant to be the type for the class, not for replacing camel case?"

The file names should actuall be kebab case I think. A great name :D 

Does it say somewhere to name the files different from classes?

Comment (antti-TIEA4) [inline app/index.html]: Directives are a bit different because the directive name is used as an element name (e.g. `<ucBackground></ucBackground>` or an attribute name inside an element (e.g. `<div ucBackground/>`). With services, factories and providers file names _can_ be named after them. But even the names in those are not exactly classes. For example the service name is an instance and the "class" (js does not really have classes) is the constructor function. In your directive case the closest to class is the anonymous function which is a factory that creates the directive. But since it is an anonymous function it doesn't have a name :) (The style guide we use actually advises not to use anonymous functions but we can look into refactoring that later after this PR) If you look at the style guide you linked to https://angular.io/styleguide#!#02-08 the blue box has a filename `validate.directive.ts` even though the directive name is `tohValidate`.
So the file and module names should not be tied to the app name prefix convention of the directive name.

Comment (heinoapt) [reply app/index.html]: Oh, well I can restrict that for the directive to work only with attributes :)

So are you saying that you can or can't name the files with the directive suffix, as you have written below with `validate.directive.ts` (which is for Angular 2)?

The filename is an non issue I think, the filenames can't cause any collisions.

My mistake on using the word "class"! I didn't think it was relevant in this conversation.

Comment (antti-TIEA4) [inline app/index.html]: I'm saying that the filenames should **not** use the directive prefix. Neither should the module (package) name. Only the directive name should use the prefix.

All the other files in this project don't have `uc` suffix, so these should not be an exception.
In other words the filenames should be:
`background.image.module.js`
`background.image.directive.js`

Comment (heinoapt) [reply app/index.html]: Yeah, what I meant that the prefix on the name is an non issue.

I still can't agree with the "image" suffix. Seems like the Style guide doesn't either:

"All other conventions still hold using a suffix of the type" and I can't find a mention of other suffixes.
- https://github.com/johnpapa/angular-styleguide/blob/master/a1/README.md#feature-file-names

Angular way seems to be kebab case for file names. So that would make it `background.directive.js` or `background-image.directive.js`, but I don't think that's so important right now. Even the "ancestor" of this directive, the HTML attribute background applies an image as the background, if such is given to it.

Comment (antti-TIEA4) [inline app/index.html]: I just thought that the word image would make it clear right from the filename that it is manipulating the background image. But I don't mind if you remove it.

> "All other conventions still hold using a suffix of the type" and I can't find a mention of other suffixes.

That part seems to say that you can drop the `controller` postfix but not with other types (such as services, providers, factories, components, directives). I think we agree on this? :)

> `background.directive.js` or `background-image.directive.js`

Both are fine with me.

Just dropping the `uc` from everywhere else except the directive name and then everything fits with the rest of the project naming.

Comment (heinoapt) [reply app/index.html]: Ok. To be honest I don't really know what you think needs to be changed anymore :D I thought that you wanted to remove the `.directive` from the file name and I couldn't really find the exact moment you changed your mind. I can't really comment on every aspect you take and I'm not really sure you would like to hear my thoughts on the wordings I use. You've mentioned quite a few details I haven't touched, but rather skimmed as irrelevant, so I might have missed some things too.

Maybe you could sum the change request?

Comment (antti-TIEA4) [inline app/index.html]: :smile: I never wanted the `.directive` postfix to be removed. Yeah, we wrote almost a bible worth of text out of a small thing, lol.

Summary:
I wanted you to drop the `uc` prefix from everywhere. I was wrong. You are right that it should be in the directive name. Everywhere else it should be removed. That's how it is in the angular.io style guide, the johnpapa style guide, and that's how all the directives and their respective files and modules are named in city_info and upcoming.

So in short:
Rename the directory `ucBackground` to `background`
Rename the file `ucBackground.module.js` to `background.module.js`
Rename the file `ucBackground.directive.js` to `background.directive.js`
Rename the module name in [this file](https://github.com/heinoapt/upcoming/blob/7dacb82248af452cafda6d23b6d5891f391560c4/app/js/shared/ucBackground/ucBackground.module.js#L5) from `'shared.ucBackground'` to `'shared.background'`

Comment (heinoapt) [reply app/index.html]: Yeah, that's something I can agree to! What I meant about the file name being a non-issue earlier, was fixing them all to that form.

Comment (heinoapt) [reply app/index.html]: These changed are now pushed.

At the same time I made a mistake with the branches, so I pushed some other stuff too. More about that on a new comment.

**Comment by heinoapt** [comment]
Pushed stuff to the event list, left navigation, by mistake.

The changes include new layout, animation and a bit of responsiveness.

File: app/index.html (hunk 1 above)
Comment (heinoapt) [reply app/index.html]: Forgot to add the directive restriction to work only with attributes. It's now pushed too.

**Comment by antti-TIEA4** [comment]
Works and now the events look really good!

Some remarks:
- Perfect scrollbar: I tested with and without it and I could not see a difference in Chrome. There's a gray scroll bar on the right and the behavior seems to be the same. If we do need this library, there might be a more Angular like way to implement it here https://github.com/itsdrewmiller/angular-perfect-scrollbar (https://github.com/itsdrewmiller/angular-perfect-scrollbar/tree/master/examples). But if tomorrow is the deadline, the current way is ok too
- Angular Routing: I don't think we are allowed to use the browser url in the LLB environment. You don't seem to be using the routing stuff yet but imported those libraries and modules?
- the $rootScope.smallEventDetails stuff could be something that the Angular responsive breakpoints are for, I have not tried them yet myself but you might want to have a look at them

Because you are working on your master branch, that's going to cause some issues when xiaodong and me introduce other changes. We can add these commits now and you can address the comments later. After we merge these you should continue on a separate branch.

**Comment by heinoapt** [comment]
Using the trunk creates problems for me too, so I would love to see it change.

I do see a difference in Chrome too, but new scroll bar doesn't show up when the list is automatically loaded. I didn't think this would be an issue because that behavior is changing soon. Just close it an open it again and you'll see a difference I believe.

Removed the route stuff, I forgot it when I tested different methods for the animation.

The method can be changed of course. If you want to include a library that's fine by me. The behavior will be the same anyway.

The

**Comment by antti-TIEA4** [comment]
Could be that importing PS this way does not make Angular aware of it in some cases.

But I'll merge this now. Make sure that you have somehow saved all works in progress because the commit history will be rewritten.

**Comment by heinoapt** [comment]
I think it's only because the scroll bar is initiated after opening it manually the first time.

Good. Thanks for the heads up. I don't think I'll run into troubles.

**Comment by heinoapt** [comment]
Just to make sure, you did see the new scroll bar and the removed referrals to the ng-route?

**Comment by heinoapt** [comment]
Works for me, the changes I've planned have been based on this review. I had to drop in the UI stuff to make my repo work.

I'll keep my hands off!

**Comment by antti-TIEA4** [comment]
In master now.

To fix your local master hard reset to a previously known clean commit (see below) and pull the changes from the upstream repo:
```
git checkout master
git reset --hard 8638e5e69e4b284e574771848b0631ac5b51df3a
git pull upstream master
```

**Comment by heinoapt** [comment]
Why is the hard reset needed?

**Comment by antti-TIEA4** [comment]
Non-hard reset keeps the reset changes in staging or in working directory and will probably cause a conflict when you pull.
Hard will throw everything away up to the given commit hash. All the changes you've made are here in this GitHub repo so you can just pull them from here.

**Comment by antti-TIEA4** [comment]
I cloned your master version from your fork and tested that the aboce commands run ok

**Comment by heinoapt** [comment]
I didn't need the hard reset on my own forked repo, just a few simple merges. You really do need to be careful with the changes if you do a hard reset.

**Comment by antti-TIEA4** [comment]
Hard reset is like a mini version of removing everything and cloning the repository again. It's ok if all the changes are in the upstream repository.
Is your local master now equal to this repository's master? If it is not, there will be issues again later. Using the GitHub UI merge to solve them might work in the short term but it will be a pain in the long term. We should all have the same commits in master as in the upstream.

**Comment by antti-TIEA4** [comment]
To fix your fork in GitHub run
```
git push --force origin master
```

**Comment by heinoapt** [comment]
Yes, I know that. That's the reason it's bad for the changes.

Is there a way to makes sure? This I just run.

```
$ git pull upstream master
From github.com:APT-project/upcoming
 * branch            master     -> FETCH_HEAD
Already up-to-date.

$ git status
On branch master
Your branch is up-to-date with 'origin/master'.
nothing to commit, working tree clean
```

**Comment by antti-TIEA4** [comment]
Your fork (and local repository on your computer)
https://github.com/heinoapt/upcoming/commits/master
should look like this
https://github.com/APT-project/upcoming/commits/master

But looks like they aren't. The state of the files might currently be the same via explicit merging with the GitHub web interface but when you push something here next time all those differing commits will be pushed with them on top of the new changes and it will mess things up. Solution is to reset to the last mutual identical commit and then pull the following commits from upstream. That way you get the exact same commits.

**Comment by heinoapt** [comment]
Ok, I can't say that I'm an expert on git, but looking at that the difference is that my repo is one commit ahead, so shouldn't that just settle down on the next PR?

**Comment by antti-TIEA4** [comment]
Look the commits before, there are around 10 commits that are not in master.

**Comment by antti-TIEA4** [comment]
And we could all have different master branches and then merge them every time we contribute. But it will introduce an unreadable commit log history (does not look good in the eyes of the person who's supposed to review this repository) and eventually it will cause all kinds of conflicts and things won't merge cleanly anymore.
I have taken that route before and would like to avoid it in the future.

**Comment by heinoapt** [comment]
Ok. So I'll just clean my repo, or in this case I'll create a new one, add a branch for that when doing changes and do a PR on that branch the APT-project-upcoming/master or somewhere else?

**Comment by antti-TIEA4** [comment]
You can use the current repository and do
```
git checkout master
git reset --hard 8638e5e69e4b284e574771848b0631ac5b51df3a
git pull upstream master
git push --force origin master
```

Or start from scratch. But I think setting everything up from scratch (you'd need to sort your GitHub fork too) might take more time.

**Comment by antti-TIEA4** [comment]
And then to introduce new changes, make sure your master is up to date with this repository (upstream) and make a new branch from that and work in that branch.
```
git checkout master
git pull upstream master
git checkout -b event_tiles
```

**Comment by antti-TIEA4** [comment]
If you had worked on a separate branch from master when making this PR you could have just removed that branch and your master would have been clean. Then you could have just pulled updates from here.
Working directly on master is a common pitfall that can cause issues like this. I fallen in that pit many times :)
